Relocate Linux Kernel to Preferred Address

phip1611 opened the pull request:
Relocate the Linux kernel to its preferred address.

phip1611 committed:
relocate Linux kernel to preferred location

phip1611 committed:
small code improvement to streamline used data types

snue commented on main.c:
```diff
@@ -137,7 +137,18 @@ static void linux_boot(const struct boot_params boot_params)
     printf("\n");
 
-    char *linux_header_end = (char *)(boot_params.kernel_addr + 0x201);
+    // If the kernel is not relocatable, we try to load Linux at the aaddress it wants to have.
+    uintptr_t actual_kernel_load_addr;
+    if (l_params->relocatable_kernel) {
+        actual_kernel_load_addr = boot_params.kernel_addr;
+        printf("Kernel is relocatable. We use the load address from the VMM\n");
+    } else {
+        actual_kernel_load_addr = l_params->pref_address;
+        printf("Kernel is not relocatable. We try the preferred load address from the kernel\n");
+    }
```
I don't think this distinction makes sense. The VMM should not force the kernel load address in any case. It's up to the loader to find a suitable place. And the "preferred address" is always the preferred address.

The VMM actually placed the kernel image at `kernel_addr`, and communicates that address via the `fw_cfg` entry. A relocatable kernel will relocate itself on boot if possible, while a non-relocatable kernel would die. Instead, the loader should move the kernel from the original `kernel_addr` to the preferred address immediately, making life easier for all bzImages. This includes checking if the current address is the preferred address (no relocation required) and if the preferred address conflicts with the kernel, initrd, or cmdline placed by the VMM (from `fw_cfg`).

The relocation should probably use `memmove` as the source and destination of the kernel might overlap.

snue commented on main.c:
```diff
@@ -137,7 +137,18 @@ static void linux_boot(const struct boot_params boot_params)
+        actual_kernel_load_addr = l_params->pref_address;
+        printf("Kernel is not relocatable. We try the preferred load address from the kernel\n");
+    }
+    printf("Loading kernel at 0x%lx\n", actual_kernel_load_addr);
```
This should say:

```suggestion
    printf("Relocating kernel from 0x%lx to 0x%lx\n", boot_params.kernel_addr, actual_kernel_load_addr);
```

snue commented on main.c:
```diff
@@ -146,18 +157,20 @@ static void linux_boot(const struct boot_params boot_params)
-    memcpy((char *)linux_params + LINUX_HEADER_OFFSET,
-           (const char *)(boot_params.kernel_addr + LINUX_HEADER_OFFSET),
+    memcpy((uint8_t *)linux_params + LINUX_HEADER_OFFSET,
+           (const uint8_t *)(actual_kernel_load_addr + LINUX_HEADER_OFFSET),
```
The second parameter to `memcpy` is the source. You did not relocate anything yet, so there is probably garbage at this address.

snue reviewed (changes requested):
This is still missing the actual relocation part :wink:

phip1611 committed:
document boot_params struct

tfc commented on main.c:
```diff
@@ -34,11 +34,16 @@ static void die_on(const bool condition, const char *string, ...)
+    // Physical address of the initrd size.
+    const uintptr_t initrd_size_addr;
+    // Physical address of the cmd line string (a null-terminated C-string).
+    const uintptr_t cmdline_addr;
```
if all these fields should be constant, why not have a const instance of `boot_params`?

phip1611 replied:
It is: `static void linux_boot(const struct boot_params boot_params)`

I have never written large C projects with others, so I do not know what modern code styles are. From my intuition, I mark everything as const because I think it is better style, isn't it?

tfc replied:
Generally yes, but while you are designing a data structure, you remove the choice from the user (maybe some test implementer or something you can't foresee) to handle the content as they wish.
While at the same time, regardless of the individual fields being const or not, the user can still always make the whole data structure const.

phip1611 replied:
thanks!

tfc commented on main.c:
```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
+        );
+
+        uintptr_t kernel_begin_addr = actual_kernel_load_addr;
+        uintptr_t kernel_end_addr = kernel_begin_addr + l_params->init_size;
```
in general more const would be nice because it reduces the amount of cognitive considerations while reading.

tfc commented on main.c:
```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
+        uintptr_t kernel_end_addr = kernel_begin_addr + l_params->init_size;
+
+        // check if the space is available in the memory map
+        for (int i = 0; i < lib_sysinfo.n_memranges; i++) {
```
the usual type for array indices is `size_t`.

tfc commented on main.c:
```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
+        }
+
+        // check that the kernel will not clash with the initrd after relocation
+        uint64_t initrd_size = *(uint64_t *)boot_params.initrd_size_addr;
```
it seems surprising/inconsistent that these fields are all `uintptr_t` but here we use `uint64_t`.

phip1611 replied:
I agree.

tfc commented on main.c:
```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
+        );
+
+        // check that the kernel will not clash with the cmdline after relocation
+        uint64_t cmdline_size = strlen((char *)boot_params.cmdline_addr);
```
use of `strlen` produces some bad feelings as there are safer alternatives.

snue replied:
Below we explicitly do not look for the actual cmdline length. There isn't even a guarantee that this is properly null terminated. We could use a reasonable constant for the maximum size here.

> The number of kernel parameters is not limited, but the length of the complete command line (parameters including spaces etc.) is limited to a fixed number of characters. This limit depends on the architecture and is between 256 and 4096 characters. It is defined in the file ./include/asm/setup.h as COMMAND_LINE_SIZE.

```
arch/x86/include/asm/setup.h:#define COMMAND_LINE_SIZE 2048
```

tfc replied:
To increase actionability: Please use `strnlen` in combination with the useful max size that @snue suggested

tfc replied:
although... if it's just for the check, using `strnlen` is overkill.

snue replied:
> although... if it's just for the check, using `strnlen` is overkill.

Yes. I recommend just checking against the constant and not even attempt to get the true command line length.

snue commented on main.c:
```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
+    uintptr_t actual_kernel_load_addr = boot_params.kernel_addr;
+
+    // Checks if a relocation of the kernel is required
+    if (boot_params.kernel_addr == actual_kernel_load_addr)
```
From the code above, this is always true.

I suggest you drop the `actual_kernel_load_addr` and work with `boot_params.kernel_addr` and `l_params->pref_address` directly. Or with our read/write copy of the [zero page](https://www.kernel.org/doc/html/latest/x86/zero-page.html) (`linux_params`) after it has been allocated and initialized.

phip1611 replied:
oops. good catch!

snue commented on main.c:
```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
+        printf("The kernel is already loaded at its preferred address.\n");
+        printf("Booting kernel at 0x%lx\n", actual_kernel_load_addr);
+    } else {
+        printf("The kernel will be moved to its preferred load address.\n");
```
This info is redundant with the following comment. We do not need to be too verbose, especially as there are no options for verbosity levels here. It should just be clear which code path was followed.

snue commented on main.c:
```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
+        for (int i = 0; i < lib_sysinfo.n_memranges; i++) {
+            die_on(kernel_begin_addr >= lib_sysinfo.memrange[i].base
+                       && kernel_end_addr < lib_sysinfo.memrange[i].base + lib_sysinfo.memrange[i].size
+                       && lib_sysinfo.memrange[i].type != CB_MEM_RAM,
```
This check is backwards. It only fails if there is a single memrange for the kernel and it is not regular memory. It won't fail if the kernel crosses memranges of which at least one is invalid.

snue reviewed (changes requested):
Needs a bit of love still.

snue commented on main.c:
```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
+            (void *) boot_params.kernel_addr,
+            (void *) (uintptr_t) l_params->pref_address,
+            l_params->init_size
+        );
```
```c
void *memmove(void *dest, const void *src, size_t n);
```

You got the parameters mixed up.

I also suggest you do the relocation as final step before jumping to the entry point. That also makes the continued use of `l_params` safer.

snue commented on main.c:
```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
+        );
+    }
+
+    uint8_t *linux_header_end = (uint8_t *)(uintptr_t)(l_params->pref_address + 0x201);
```
The `l_params` point to within the original location of the kernel image. It should not be used after the `memmove` since with overlapping kernel locations this might not be the original contents anymore.

Allocate the `linux_params` before the relocation and check it also does not collide with the intended kernel space. Then always use `linux_params` instead of `l_params`.

snue commented on main.c:
```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
+        printf(
+            "The kernel is already loaded at its preferred address 0x%llx\n",
+            l_params->pref_address
+        );
```
nit: I find this too chatty. I'd rather not print anything for this case.

snue commented on main.c:
```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
+        );
+    } else {
+        printf(
+            "The kernel will be moved from 0x%lx to its preferred load address 0x%llx.\n",
```
nit: why so much text? What's wrong with:

```suggestion
            "Relocating kernel from 0x%lx to 0x%llx.\n",
```

snue commented on main.c:
```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
+            l_params->pref_address
+        );
+
+        uintptr_t kernel_begin_addr = l_params->pref_address;
```
I just realized these fields were only introduced with boot protocol version `2.10`:
https://www.kernel.org/doc/html/latest/x86/boot.html

The version check should be adapted accordingly.

snue commented on main.c:
```diff
@@ -146,18 +212,17 @@ static void linux_boot(const struct boot_params boot_params)
-    memcpy((char *)linux_params + LINUX_HEADER_OFFSET,
-           (const char *)(boot_params.kernel_addr + LINUX_HEADER_OFFSET),
+    memcpy((uint8_t *)linux_params + LINUX_HEADER_OFFSET,
+           (const uint8_t *)(uintptr_t)(l_params->pref_address + LINUX_HEADER_OFFSET),
```
why the double cast?

phip1611 replied:
otherwise there is a compiler error

blitz replied:
To avoid ` warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]`. In this case, it would be harmless, but there is some sadness regarding sign-extension that make some of these conversions highly unintuitive...

snue commented on main.c:
```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
+
+        // check that the kernel will not clash with the cmdline after relocation
+        // value taken from: arch/x86/include/asm/setup.h:#define COMMAND_LINE_SIZE 2048
+        const size_t MAX_LEN = 2048;
+        uintptr_t cmdline_size = strnlen((char *)boot_params.cmdline_addr, MAX_LEN);
```
The constant can be moved to some header file. Maybe the `linux_params.h`.

Also please just check against the constant. Determining the actual cmdline length is overkill.

snue replied:
Actually, since we already rely on protocol version 2.10 now, we can just take the `cmdline_size`, which is included since boot protocol version 2.06.

snue commented on main.c:
```diff
@@ -146,18 +212,17 @@ static void linux_boot(const struct boot_params boot_params)
            linux_header_size); // load header
 
     die_on(linux_params->param_block_version < 0x0205,
            "Kernel does not support boot protocol >= 2.05\n");
```
Here, this needs to be `2.10` if we rely on `pref_address`.

snue commented on main.c:
```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
+
+        // check that the kernel will not clash with the initrd after relocation
+        uintptr_t initrd_size = *(uint64_t *)boot_params.initrd_size_addr;
+        die_on(kernel_begin_addr >= boot_params.initrd_addr
+                   && l_params->pref_address < boot_params.initrd_addr + initrd_size,
+               "Kernel load address collides with initrd\n"
+        );
```
If we ever run into this, we could try to relocate the initrd as well. But that deteriorates into quite some memory area juggling.

phip1611 replied:
maybe, we should do this in another MR?

snue replied:
Definitley nothing to be done right now. The assumption should also be that the ramdisk is basically loaded to the highest allowed address and shouldn't conflict with any typical kernel load address.

snue reviewed (changes requested):
Found more things to address.

phip1611 committed:
Require boot protocol 2.10

We require Linux boot protocol version 2.10 and above from
now on because we want to read the preferred address from
the kernel image and load kernels to this address.

phip1611 committed:
more early checks & logs: verify Linux image signature

phip1611 committed:
linux_boot(): use const where possible
